### PR TITLE
Feat/keypoints from mediapipe

### PR DESCRIPTION
**Pull request by David-rn**
# Description

According to issue #1174, this PR extends the usage of KeyPoints class to add `from_mediapipe` connector. 

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

The following [Google Colab](https://colab.research.google.com/drive/1XN5yuqNDtS8Rakfh3q_Qxq_oo-WGfR0A?usp=sharing) describes the process used for testing this new feature. It allows to upload a picture to test it. (This Colab feature only worked in Chrome for me)

## Any specific deployment considerations

- This connector needs an additional parameter `resolution` since the result from mediapipe only provides the normalized coordinates [0.0, 1.0]. I didn't find the equivalent of [VideoInfo](https://github.com/roboflow/supervision/blob/2723735b018cf6f8907d4c706b5f759ff641a53d/supervision/utils/video.py#L15) for images, so the parameter accepts a Tuple of ints. 

- Another consideration is that mediapipe only outputs the prediction of [the most prominent person](https://github.com/google-ai-edge/mediapipe/blob/master/docs/solutions/pose.md) in the image. So, even if there is several people in the image, the output will be only the keypoints from one person. The implementation has taken this into account.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

**Comment by SkalskiP** [comment]
Hi, @David-rn 👋🏻 Thanks a lot for submitting your PR. @LinasKo, could you take a look?

File: supervision/keypoint/core.py [hunk 1]
```diff
@@ -185,6 +186,57 @@ class IDs, and confidences of each keypoint.
...
+    ) -> KeyPoints:
+        """
+        Creates a KeyPoints instance from a
+            [Pose landmark detection](https://ai.google.dev/edge/mediapipe/)
+                inference result.
+
+        Args:
+            mediapipe_results (mediapipe.python.solution_base.SolutionOutputs):
+                The output Results from Mediapipe
+            resolution (Tuple[int, int]): image resolution (w, h) since mediapipe
```
[the remark this answers is not in the document]
Comment (LinasKo) [reply supervision/keypoint/core.py:201]: I like the explanation for why it's needed :)

For our reference: `pose_world_landmarks` also normalized coordinates, but adjusted so origin is the point between the hips of the person.

File: supervision/keypoint/core.py [hunk 2]
```diff
@@ -185,6 +186,57 @@ class IDs, and confidences of each keypoint.
...
+            prediction_xy.append(
+                [
+                    min(math.floor(landmark.x * resolution[0]), resolution[0] - 1),
+                    min(math.floor(landmark.y * resolution[1]), resolution[1] - 1),
+                ]
+            )
+            prediction_confidence.append(landmark.visibility)
+
+        return cls(
+            xy=np.array([prediction_xy], dtype=np.float32),
```
Comment (LinasKo) [inline supervision/keypoint/core.py:236]: Let's make sure that when no predictions are returned, the empty array is still shaped `(0, -2)`.

For example:

```
if len(prediction_xy) == 0:
  return cls(
    xy=np.empty((1, 0, 2), dtype=np.float32),
    confidence=np.empty((1, 0,), dtype=np.float32)
)
```

**Review by LinasKo — CHANGES_REQUESTED** [review]
Overall, this is a very neat addition to our codebase. Easy to test, immediately visible benefit, neatly coded-up.

I left a few smaller changes, and I'd like to ask for some larger ones:

1. I'd like to verify that using this inside a `PoseLandmarker` context block works. If you follow the [guide](https://ai.google.dev/edge/mediapipe/solutions/vision/pose_landmarker/python) you'll see `PoseLandmarkerOptions` being created and then `with PoseLandmarker.create_from_options(options) as landmarker:` called. Inside, `Detections.from_mediapipe` should work.
2. Next, and that's something I've overlooked for the longest time didn't - MediaPipe does in fact support more than 1 person! To do it, you need to set the number of poses explicitly by passing `num_poses=N` to `PoseLandmarkerOptions`.

Would you mind making sure these work and adding the test to the Colab?
The docstring `from_mediapipe` doesn't need to mentioned these examples.

**Comment by David-rn** [comment]
Hi @LinasKo! Thanks for the suggestions!

> 1. I'd like to verify that using this inside a `PoseLandmarker` context block works. If you follow the [guide](https://ai.google.dev/edge/mediapipe/solutions/vision/pose_landmarker/python) you'll see `PoseLandmarkerOptions` being created and then `with PoseLandmarker.create_from_options(options) as landmarker:` called. Inside, `Detections.from_mediapipe` should work.

Regarding the first comment I discover that the use of `mp.solutions.pose.Pose` (the one I implemented) is a Legacy Solution as of March 1, 2023 as state [here](https://ai.google.dev/edge/mediapipe/solutions/guide). This confusion originated in the example Colab. The one you suggest in your comment is the new way to go. However, the result object is not the same as the legacy one. 

For this reason, should we provide support for both types or just focus on the new result? This is quite linked to the next comment.

> 2. Next, and that's something I've overlooked for the longest time didn't - MediaPipe does in fact support more than 1 person! To do it, you need to set the number of poses explicitly by passing `num_poses=N` to `PoseLandmarkerOptions`.

Regarding this second comment, indeed, it is now possible to obtain more than one pose, but only with the new implementation whose output object is of type [`PoseLandmarkerResult`](https://github.com/google-ai-edge/mediapipe/blob/2da0c56a5df3463af869dc365339225f0553d43b/mediapipe/tasks/python/vision/pose_landmarker.py#L57). This object holds a list of poses whereas the legacy one holds a different structure.

**Comment by LinasKo** [comment]
1. How hard would it be to support both? If it's tricky to reason about or the code involved is complicated, let's stick to the new version. Even a few years back, I was already using the newer API, I believe.
2. Aye, let's add the newer one for sure in that case.

**Comment by David-rn** [comment]
I think it wasn't that hard, but let me know if it is a nice solution:

I parsed the legacy result into a list (same as the new one) so the loop to fill the new KeyPoints object can remain the same for both results. I also updated the [Google Colab](https://colab.research.google.com/drive/1XN5yuqNDtS8Rakfh3q_Qxq_oo-WGfR0A?usp=sharing) to make easier the checking process.

**Comment by LinasKo** [comment]
Hi @David-rn,

Apologies for the delay! Good to hear it supports both cases - I'm prioritizing to review this next week.

**Comment by LinasKo** [comment]
Tidied up a few small aspects of this:
1. `(` instead of `[` making `mypy` happy
2. Removed legacy example. With our current plans to also add support for mediapipe face, this would bloat the docs.
3. Brought up to speed with `develop`.
4. Added inline suggestion to download model task. Initially intended to add a URL outside the code block, but it doesn't render nicely.
5. Tested on a local webcam.

<details>
  <summary>Code</summary>
  
  ```python
  import mediapipe as mp  # noqa
  import cv2
  import numpy as np
  import supervision as sv
  
  
  BaseOptions = mp.tasks.BaseOptions
  PoseLandmarker = mp.tasks.vision.PoseLandmarker
  PoseLandmarkerOptions = mp.tasks.vision.PoseLandmarkerOptions
  VisionRunningMode = mp.tasks.vision.RunningMode
  
  options = PoseLandmarkerOptions(
      base_options=BaseOptions(model_asset_path="pose_landmarker_lite.task"),
      running_mode=VisionRunningMode.IMAGE,
      num_poses=4,
  )
  
  cap = cv2.VideoCapture(0)
  
  
  with PoseLandmarker.create_from_options(options) as landmarker:
      # pose_landmarker_result = landmarker.detect(mp_image)
      while cap.isOpened():
          ret, frame = cap.read()
          if not ret:
              break
          mp_image = mp.Image(
              image_format=mp.ImageFormat.SRGB,
              data=cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
          )
  
          pose_landmarker_result = landmarker.detect(mp_image)
  
          image_hight, image_width, _ = frame.shape
          keypoints = sv.KeyPoints.from_mediapipe(pose_landmarker_result, (image_width, image_hight))
  
          edge_annotator = sv.EdgeAnnotator(color=sv.Color.GREEN, thickness=2)
          annotated_image = edge_annotator.annotate(
              scene=frame.copy(),
              key_points=keypoints,
          )
  
          cv2.imshow("image", annotated_image)
          if cv2.waitKey(1) & 0xFF == ord('q'):
              break
          
  cap.release()

  ```
</details>

I like how this looks. @SkalskiP, up for giving the code a quick glance?

**Review by LinasKo — APPROVED** [review]

**Pull request merged**